File: src/main/java/com/mercateo/jsonschema/generictype/GenericTypeAbstract.java
package com.mercateo.jsonschema.generictype;

import java.lang.reflect.Field;
import java.lang.reflect.Type;
import java.util.Objects;

import static java.util.Objects.requireNonNull;

abstract class GenericTypeAbstract<T, U extends Type> implements GenericType<T> {

    protected final Class<T> rawType;

    protected final U type;

    GenericTypeAbstract(Class<T> rawType, U type) {
        this.rawType = requireNonNull(rawType);
        this.type = type;
    }

    @Override
    public final Class<T> getRawType() {
        return rawType;
    }

    @Override
    public U getType() {
        return type;
    }

    @Override
    public boolean isInstanceOf(Class<?> clazz) {
        return requireNonNull(clazz).isAssignableFrom(getRawType());
    }

    @Override
    public boolean isIterable() {
        return Iterable.class.isAssignableFrom(getRawType());
    }

    @Override
    public Field[] getDeclaredFields() {
        return getRawType().getDeclaredFields();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || !GenericType.class.isAssignableFrom(o.getClass())) {
            return false;
        }
        if (getClass() != o.getClass()) return false;
        GenericType<?> that = (GenericType<?>) o;
        return Objects.equals(rawType, that.getRawType()) &&
            Objects.equals(type, that.getType());
    }

    @Override
    public int hashCode() {
        return Objects.hash(type, type);
    }
}

File: src/main/java/com/mercateo/jsonschema/generictype/GenericTypeAbstract.kt
package com.mercateo.jsonschema.generictype

import com.googlecode.gentyref.GenericTypeReflector
import java.lang.reflect.Field
import java.lang.reflect.Method
import java.lang.reflect.Type
import java.util.*
import java.util.Objects.requireNonNull

internal abstract class GenericTypeAbstract<out T, out U : Type>(
        override val rawType: Class<out T>,
        override val type: U
) : GenericType<T> {

    override val simpleName: String
        get() = rawType.simpleName

    override val name: String
        get() = rawType.name

    override fun isInstanceOf(clazz: Class<*>): Boolean {
        return requireNonNull(clazz).isAssignableFrom(rawType)
    }

    override val isIterable: Boolean
        get() = Iterable::class.java.isAssignableFrom(rawType)

    override val isEnum: Boolean
        get() = rawType.isEnum

    override val declaredFields: Array<Field>
        get() = rawType.declaredFields

    override val declaredMethods: Array<Method>
        get() = rawType.declaredMethods

    override val superType: GenericType<Any>?
        get() {
            val superclass = rawType.superclass as Class<out Any>?
            return if (superclass != null) {
                val exactSuperType: Type = GenericTypeReflector.getExactSuperType(
                        type, superclass)
                GenericType.of(exactSuperType, superclass)
            } else {
                null
            }
        }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || !GenericType::class.java.isAssignableFrom(other.javaClass)) {
            return false
        }
        if (javaClass != other.javaClass) return false
        val that = other as GenericType<*>?
        return rawType == that!!.rawType && type == that.type
    }

    override fun hashCode(): Int {
        return Objects.hash(type, type)
    }
}

